### PR TITLE
Fix GRPO use_liger_kernel under DeepSpeed ZeRO-3

### LINKED ISSUE #3368: GPRO: use_liger_loss + zero3 error
### Reproduction

✅ use_liger_loss=true + zero2
✅ use_liger_loss=false + zero3
❌ use_liger_loss=true + zero3

error：size mismatch

### System Info

pytorch=2.6
deepspeed=0.16.4

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by kashif**
## What this fixes

`use_liger_kernel=True` + DeepSpeed ZeRO-3 in GRPO crashes with a size mismatch. The matrix below is what people have been hitting (see #3368):

- ✅ liger + zero2
- ✅ no-liger + zero3
- ❌ liger + zero3 → size mismatch

## Why

In `compute_liger_loss` we hand `unwrapped_model.lm_head.weight` straight to the Liger fused loss. Under ZeRO-3 that weight is sharded across ranks, and since Liger reads it by attribute access instead of going through `model.forward()`, DeepSpeed's gather hook never fires. So on the partitioned ranks the fused matmul gets an empty shard and blows up.

## Fix

Wrap the Liger call in a `GatheredParameters` context so the full `lm_head` weight/bias are all-gathered for the matmul, then re-partitioned on exit. This is the same `_maybe_gather_lm_head_ctx` pattern we use for SFT's chunked-NLL path. It's a no-op when not on ZeRO-3, or when the weight is already gathered (tied embeddings keep it `AVAILABLE`).

We only need the gathered weight during the forward: `LigerFusedLinearGRPOLoss` computes the gradient w.r.t. the weight eagerly inside the forward and saves it for backward, so re-partitioning before `backward()` is fine.

## Test

Added `test_grpo_liger` to the distributed suite, parametrized over ddp / zero2 / zero3 / fsdp2 with `--use_liger_kernel` (mirrors the existing `test_grpo`).

Closes #3368

---

> [!NOTE]
> **Medium Risk**
> Touches GRPO training loss under ZeRO-3 and tied-embedding edge cases; scope is narrow and covered by new distributed tests.
> 
> **Overview**
> Fixes a crash when **GRPO** runs with **`use_liger_kernel`** under **DeepSpeed ZeRO-3**: Liger reads `lm_head` weights directly, so ZeRO-3 never gathers shards and the fused matmul fails with a size mismatch.
> 
> In **`compute_liger_loss`**, the Liger call is wrapped in **`deepspeed.zero.GatheredParameters`** for `lm_head` weight/bias when ZeRO-3 is active and params are not already `AVAILABLE` (e.g. tied embeddings). Otherwise behavior is unchanged.
> 
> Adds distributed **`test_grpo_liger`** (ddp / zero2 / zero3 / fsdp2) mirroring **`test_grpo`** with **`--use_liger_kernel`**, gated by **`require_liger_kernel`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b4d345ed1f18e31173889e345b7b6b9aaadf7a11. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by kashif** [comment]
@codex review

File: tests/distributed/test_distributed.py [hunk 1]
```diff
@@ -301,3 +301,43 @@ def test_grpo(self, config, get_config_path):
...
+                marks=pytest.mark.xfail(
+                    Version("5.0.0") <= Version(transformers.__version__) < Version("5.5.4"),
+                    reason="ZeRO-3 fails with transformers >= 5.0.0 and < 5.5.4 (fixed in transformers#45414), see #4899",
+                    strict=True,
+                ),
+            ),
+            "fsdp2",
+        ],
+    )
+    def test_grpo_liger(self, config, get_config_path):
```
[the remark this answers is not in the document]
Comment (kashif) [reply tests/distributed/test_distributed.py:328]: Addressed in 8cfbdb1d6 — added `@require_liger_kernel` so this skips cleanly when the optional extra isn't installed, matching the other Liger tests.

[the remark this answers is not in the document]
Comment (kashif) [reply tests/distributed/test_distributed.py:328]: Addressed in 8cfbdb1d6 — added `@require_liger_kernel` so this skips cleanly when the optional extra isn't installed, matching the other Liger tests.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -2325,19 +2325,35 @@ def compute_liger_loss(self, unwrapped_model, inputs):
...
+        gather_ctx = nullcontext()
+        if deepspeed_plugin is not None and deepspeed_plugin.zero_stage == 3:
+            from deepspeed.runtime.zero.partition_parameters import ZeroParamStatus
+
+            params = [lm_head_weight] if lm_head_bias is None else [lm_head_weight, lm_head_bias]
+            if any(p.ds_status != ZeroParamStatus.AVAILABLE for p in params):
+                import deepspeed
+
+                gather_ctx = deepspeed.zero.GatheredParameters(params)
+        with gather_ctx:
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/grpo_trainer.py:2344]: Good catch — fixed in b4d345ed1. This is a read-only gather (Liger reads `lm_head` for the matmul; the weight grad flows back through autograd, not into the param data), so `modifier_rank=None` is correct and avoids the unnecessary exit broadcast. Also matches the existing read-only Liger gather sites in the distillation/gold trainers.

**Review by qgallouedec — APPROVED** [review]
thanks!
